Add Boba migration tool (boba-chain-ops)

boyuan-chen opened the pull request:
Boba migration tool is based on op-chain-ops

It does serval different things.
1. It reads allocation data from the json file, modifies the contract bytecode, and adds proxy contracts
2. The genesis block is created based on the modified allocation data
3. It moves ETH balance from the storage of OVM_ETH contract to the state

boyuan-chen committed:
Initial commit

jyellick commented on boba-chain-ops/crossdomain/precheck.go:
```diff
@@ -0,0 +1,124 @@
+	"errors"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/log"
```
Probably doesn't really matter, but, I'd think we should be using the ledgerwatch log package here given that the other imports are erigon.

boyuan-chen replied:
Yes! Fixed.

jyellick commented on boba-chain-ops/deployer/deployer.go:
```diff
@@ -0,0 +1,129 @@
+		GrayGlacierBlock:      big.NewInt(0),
+		// Activated proof of stake. We manually build/commit blocks in the simulator anyway,
+		// and the timestamp verification of PoS is not against the wallclock,
+		// preventing blocks from getting stuck temporarily in the future-blocks queue, decreasing setup time a lot.
```
These lines have been commented out relative to the original deployer.go -- I _think_ this is covered by the changes above, but, as a nit, might be worth either deleting and removing the comment above, or, adding to the comment saying it's geth specific and covered above.

boyuan-chen replied:
Removed the comment

jyellick commented on boba-chain-ops/ether/migrate.go:
```diff
@@ -0,0 +1,299 @@
+	errCh := make(chan error, len(storage))
+
+	for key, value := range storage {
+		go cb(key, value, errCh)
```
Nothing strictly wrong with this approach, but, I think it would be cleaner if we were to eliminate the errCh construct and simply use the sync/errgroup package.  We could leave the original return pattern of directly returning errors and not dealing with the channel logic.

Something like:

```
var eg errgroup.Group
for key, value := range storage {
    eg.Go(cb)
}
if err := eg.Wait(); err != nil {
    return errr
}
```

boyuan-chen replied:
Yes, the channel is removed and replaced by `errgroup`

jyellick commented on boba-chain-ops/ether/migrate_test.go:
```diff
@@ -0,0 +1,425 @@
+		CalcOVMETHStorageKey(common.Address{2}):                       {2},
+		CalcAllowanceStorageKey(common.Address{1}, common.Address{2}): {3},
+	}
+	if !reflect.DeepEqual(accountState, g.Alloc[common.Address{1}]) {
```
Given that we're already using testify through much of the rest of the code, I think it would make more sense for us to use a:

```
require.Equal(t, accountState, g.Alloc[common.Address{1}])
```

This will eliminate the need to try to do our own formatting on not equal, and, it will more reliably handle some cases where reflect.DeepEqual isn't actually what you want.  (This comment holds through the tests)

boyuan-chen replied:
Sounds good.

jyellick commented on boba-chain-ops/genesis/alloc_migration_test.go:
```diff
@@ -0,0 +1,60 @@
+
+import (
+	"encoding/json"
+	"reflect"
```
Same note here about not using `reflect` for assertions in tests, let's prefer testify (or some other matcher like gomega if you prefer)

jyellick commented on boba-chain-ops/genesis/check.go:
```diff
@@ -0,0 +1,810 @@
+	"math/big"
+	"math/rand"
+
+	"github.com/ethereum/go-ethereum/log"
```
Another instance where we should probably be using ledgerwatch/log

jyellick commented on boba-chain-ops/genesis/db_migration.go:
```diff
@@ -0,0 +1,279 @@
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/log"
```
Last one of these I'll note -- probably worth grepping for.

jyellick reviewed (commented):
So, this is obviously a pretty big PR, and there's lots of logic and it's a bit hard for me to wrap my head around it all.  A few high level questions:

1. Withdrawals: The optimism variant has commands related to processing withdrawals and a lot of logic around how to handle pending withdrawals.  We don't seem to have preserved this.  Could you explain what the optimism stuff is doing and why we don't need it?

2. Optimism also has an EoF crawler and related code.  Same question, do you know what it's for and why we don't need it?

3. Could you add/include some documentation in a README.md as to the high level flow of what's happening? There was one in the original dir that I believe was deleted.

Finally, I've added a number of comments.  Sorry if some of them come across as bike-shedding -- but hopefully there's some useful in there.

Boyuan-Chen commented:
> So, this is obviously a pretty big PR, and there's lots of logic and it's a bit hard for me to wrap my head around it all. A few high level questions:
> 
> 1. Withdrawals: The optimism variant has commands related to processing withdrawals and a lot of logic around how to handle pending withdrawals.  We don't seem to have preserved this.  Could you explain what the optimism stuff is doing and why we don't need it?
> 2. Optimism also has an EoF crawler and related code.  Same question, do you know what it's for and why we don't need it?
> 3. Could you add/include some documentation in a README.md as to the high level flow of what's happening? There was one in the original dir that I believe was deleted.
> 
> Finally, I've added a number of comments. Sorry if some of them come across as bike-shedding -- but hopefully there's some useful in there.

I will add README to explain the whole package in a high level.
1. I am not sure about it. It's related to the L1 smart contract, but it has nothing to do with L2. We should find it out after digging into smart contracts.
2. Not sure. I didn't see that the EOF output is used in the migration

boyuan-chen committed:
Fix import root

boyuan-chen committed:
Remove channel and use erigon log

boyuan-chen committed:
Add boba predeployed contracts

boyuan-chen committed:
Fix typo and a bug

boyuan-chen committed:
Add ETH balance tests for rollover and fix extra data

boyuan-chen committed:
Support the latest op-erigon

boyuan-chen committed:
Fix difficulty of genesis block

boyuan-chen committed:
Use engine api to regenerate the chain

boyuan-chen committed:
Add docker file and fix tests

boyuan-chen committed:
Apply tests using counterfeiter

boyuan-chen committed:
Add boba crawler

boyuan-chen committed:
Simplify input

boyuan-chen committed:
Fix Dockerfile and docker-compose file

boyuan-chen committed:
Fix typo

boyuan-chen committed:
Add script to generate genesis file for l2 and add l1 boba token address as an input

boyuan-chen committed:
Add README

boyuan-chen committed:
Fix l1BobaAddress and add rollup config

boyuan-chen committed:
Fix error check

boyuan-chen committed:
Remove some require.Error

boyuan-chen committed:
Add documentation for starting a local node

boyuan-chen committed:
Merge pull request #10 from bobanetwork/boba-migration-devnet

Add script to generate genesis file for devnet

boyuan-chen committed:
Merge pull request #9 from bobanetwork/boba-migration-engine-api

Add tool to regenerate the chain using engine API

boyuan-chen committed:
Merge pull request #8 from bobanetwork/boba-migration-boba-contracts

Add Boba predeployed contracts

boyuan-chen committed:
Add tool to create empty block

boyuan-chen committed:
Add tests for a transition block

boyuan-chen committed:
Store rollup configuration

boyuan-chen committed:
Fix the default of deployment addresses

boyuan-chen committed:
Add allocation dump script

boyuan-chen committed:
Fix lints and tests

boyuan-chen committed:
Add artifacts for testing purpose

boyuan-chen committed:
Fix boba chain ops test

boyuan-chen committed:
Fix contract lints and avoid failure tests

boyuan-chen committed:
Merge branch 'boba-develop' into boba-migration

mmontour1306 committed:
Add definitions for Boba mainnet
 Changes to be committed:
	modified:   chain/chain.go
	modified:   crossdomain/params.go

boyuan-chen committed:
Add migration doc

boyuan-chen committed:
Fix genesis information for Boba Mainnet

boyuan-chen committed:
Fix typo

mmontour1306 committed:
Add a script to dump a partial block history from a temporary
erigon DB and load it into the migration target to reduce
the number of blocks which must be copied from the legacy system.

 Changes to be committed:
	new file:   db-history/dump-load-history.sh

boyuan-chen committed:
Merge pull request #14 from bobanetwork/mm/boba-migration-db-history

Add a script to dump a partial block history from a temporary DB

boyuan-chen committed:
Standardize variable names

boyuan-chen committed:
Merge pull request #13 from bobanetwork/mm/boba-migration-mainnet

Add definitions for Boba mainnet

boyuan-chen committed:
Adjust gas limit in test

boyuan-chen committed:
Adjust test order

boyuan-chen committed:
Fix contract lint

boyuan-chen committed:
Fix integration tests

boyuan-chen committed:
Update gas usage for snapshot

boyuan-chen committed:
Disable check in boba-chain-ops

boyuan-chen committed:
Disable checks of boba chain ops

boyuan-chen committed:
Fix tests

boyuan-chen committed:
Fix parallel tests

boyuan-chen committed:
Lock parallel tests

boyuan-chen committed:
Set noCheck to true

boyuan-chen committed:
Add Bobaopera configuration

boyuan-chen committed:
Merge branch 'boba-develop' into boba-migration

boyuan-chen committed:
Add postinstall

boyuan-chen committed:
Update pnpm-lock.yaml

boyuan-chen committed:
Fix contract tests

jyellick commented on boba-bindings/Makefile:
```diff
@@ -0,0 +1,154 @@
+SHELL := /bin/bash
```
This has diverged quite a bit from upstream at this point.  Might be worth adopting what's been done there?

boyuan-chen replied:
The Makerfile is synced to op-bindings.

jyellick commented on .circleci/config.yml:
```diff
@@ -1426,6 +1426,18 @@ workflows:
       - fuzz-op-node
       - fuzz-op-chain-ops
       - bedrock-markdown
+      - go-lint:
```
The upstream tests for the bindings generate the bindings, and make sure that there is no diff based on what's in the source tree -- I think this is a good idea.  Can we add there here too?

boyuan-chen replied:
This requires us to build a specific image for the testing purpose, since we use the abigen from erigon. Also, some smart contracts import the same openzepplin's smart contract, so some generated variables separate among serval files. We can only keep one of them.

boyuan-chen replied:
The test is added. https://app.circleci.com/pipelines/github/bobanetwork/v3-anchorage/164/workflows/82ad7198-25d8-4b1a-9f5c-43a596a1070f/jobs/9710

jyellick commented on boba-bindings/bindings/registry.go:
```diff
@@ -0,0 +1,39 @@
+package bindings
```
Having spent a bit more time here lately, I'm starting to understand what's going on a bit better.

As best as I can tell, the `_more.go` files bind into these variables like `layouts` files, and are generated.

It's a bit weird to me, that this file is not generated, while the rest of the files in this binding directory are.  But, I guess this is not our pattern.

Still, it might be helpful to describe exactly what's going on here in the README

boyuan-chen replied:
Added

jyellick commented on boba-bindings/gen_bindings.sh:
```diff
@@ -0,0 +1,59 @@
+#/bin/bash
```
Same here, this bash has disappeared from upstream, so maybe good to move to the newer variant.

jyellick reviewed (commented):
This is large enough, that I think I'm going to have to finish reviewing tomorrow.  The biggest overall question/concern would be that we've copied much of the code from the upstream corresponding op-* directories, and that code has already drifted quite a bit.  Do we want to sync before merging this? Or, if not, have we audited to make sure that at least there are no bug fixes that have gone in?

Boyuan-Chen commented:
> This is large enough, that I think I'm going to have to finish reviewing tomorrow. The biggest overall question/concern would be that we've copied much of the code from the upstream corresponding op-* directories, and that code has already drifted quite a bit. Do we want to sync before merging this? Or, if not, have we audited to make sure that at least there are no bug fixes that have gone in?

Sounds good! I will sync the op-bindings.

boyuan-chen committed:
Sync op-bindings

boyuan-chen committed:
Fix abigen directory

boyuan-chen committed:
Add boba-bindings check

Commit:
Update bindings

boyuan-chen committed:
Update bytecodes

boyuan-chen committed:
Update README

boyuan-chen committed:
Merge branch 'boba-develop' into boba-migration
